src/aws-cpp-sdk-core/source/client/AWSClient.cpp

```diff
@@ -874,24 +877,53 @@ Aws::String Aws::Client::GetAuthorizationHeader(const Aws::Http::HttpRequest& ht
     return authHeader.substr(signaturePosition + strlen(Aws::Auth::SIGNATURE) + 1);
 }
 
-void AWSClient::BuildHttpRequest(const Aws::AmazonWebServiceRequest& request,
-    const std::shared_ptr<HttpRequest>& httpRequest) const
+void AWSClient::BuildHttpRequest(const Aws::AmazonWebServiceRequest& request, const std::shared_ptr<HttpRequest>& httpRequest) const
 {
-    //do headers first since the request likely will set content-length as it's own header.
+    //do headers first since the request likely will set content-length as its own header.
     AddHeadersToRequest(httpRequest, request.GetHeaders());
     AddHeadersToRequest(httpRequest, request.GetAdditionalCustomHeaders());
 
-    AddChecksumToRequest(httpRequest, request);
-
     if (request.IsEventStreamRequest())
     {
         httpRequest->AddContentBody(request.GetBody());
     }
     else
     {
-        AddContentBodyToRequest(httpRequest, request.GetBody(), request.ShouldComputeContentMd5(), request.IsStreaming() && request.IsChunked() && m_httpClient->SupportsChunkedTransferEncoding());
+        //Check if compression is required
+        if (Aws::Client::CompressionAlgorithm::NONE !=
+            request.GetSelectedCompressionAlgorithm(m_requestCompressionConfig)) {
+            Aws::Client::RequestCompression rc;
+            auto compressOutcome = rc.compress(request.GetBody(), request.GetSelectedCompressionAlgorithm(m_requestCompressionConfig));
+
+            if (compressOutcome.IsSuccess()) {
+                Aws::String compressionAlgorithmId =
+                    Aws::Client::GetCompressionAlgorithmId(
+                      request.GetSelectedCompressionAlgorithm(m_requestCompressionConfig));
+
+                // Update/Add Header for compressionAlgorithmId
+                if (!httpRequest->HasHeader(CONTENT_ENCODING_HEADER)) {
+                    httpRequest->SetHeaderValue(CONTENT_ENCODING_HEADER, compressionAlgorithmId);
+                } else {
+                    Aws::String contentEncoding = httpRequest->GetHeaderValue(CONTENT_ENCODING_HEADER);
+                    contentEncoding.append(",").append(compressionAlgorithmId);
+                    httpRequest->SetHeaderValue(CONTENT_ENCODING_HEADER, contentEncoding);
+                }
```

Comment:
It worth extracting this block into a (possibly inline) method on HttpRequest with a name such as "AppendHeaderValue".

Reply:
extracted into AwsClient to keep it consistent with AddHeaders and AddBody